app/api/dds/v1/trashbin_api.rb

```diff
@@ -0,0 +1,134 @@
+          [404, 'Object kind not supported']
+        ]
+      end
+      params do
+        requires :object_kind, type: String, desc: 'Object Kind'
+        requires :object_id, type: String, desc: 'Object UUID'
+      end
+      get '/trashbin/:object_kind/:object_id', root: false do
+        authenticate!
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        trashed_object = object_kind.find_by!(id: params[:object_id], is_deleted: true, is_purged: false)
+        authorize trashed_object, :show?
+        trashed_object
+      end
+
+      desc 'Restore a Trashbin Item' do
+        detail 'Restores the item, and any children, to an undeleted status to the specified parent folder or project.'
+        named 'restore trashbin item'
+        failure [
+          [200, 'Success'],
+          [404, 'Parent object does not exist or is itself in the trashbin'],
+          [401, 'Unauthorized'],
+          [403, 'Forbidden'],
+          [404, 'Object or Parent kind not supported'],
+          [404, 'Object not found in trash bin'],
+          [404, 'Parent object does not exist or is itself in the trashbin']
+        ]
+      end
+      params do
+        requires :object_kind, type: String, desc: 'Object Kind'
+        requires :object_id, type: String, desc: 'Object UUID'
+        optional :parent, desc: "Parent. Required unless restoring a FileVersion", type: Hash do
+          optional :kind, type: String, desc: 'Parent Kind'
+          optional :id, type: String, desc: 'Parent UUID'
+        end
+      end
+      rescue_from TrashbinParentException do |e|
+        message, suggestion = e.message.split('::')
+        error_json = {
+          "error" => "404",
+          "code" => "not_provided",
+          "reason" => message,
+          "suggestion" => suggestion
+        }
+        error!(error_json, 404)
+      end
+      rescue_from IncompatibleParentException do |e|
+        message, suggestion = e.message.split('::')
+        error_json = {
+          "error" => "404",
+          "code" => "not_provided",
+          "reason" => message,
+          "suggestion" => suggestion
+        }
+        error!(error_json, 404)
+      end
+      rescue_from UnRestorableException do |e|
+        error_json = {
+          "error" => "404",
+          "code" => "not_provided",
+          "reason" => "#{e.message} Not Restorable",
+          "suggestion" => "#{e.message} is not Restorable"
+        }
+        error!(error_json, 404)
+      end
+      put '/trashbin/:object_kind/:object_id/restore', root: false do
+        authenticate!
+        parent_params = declared(params, {include_missing: false}, [:parent])
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        purge_object = object_kind.find_by!(id: params[:object_id], is_deleted: true)
+        raise UnRestorableException.new(purge_object.kind) unless purge_object.class.include? Restorable
+        unless purge_object.is_purged?
```

Comment:
I still don't know what to think about checking `.is_purged?` when restoring and not throwing an error or something. If it's gotten this far, `purge_object` is `is_deleted: true`, so if it's also `.is_purged?`, then trying to restore it shouldn't work, right?

Reply:
I think I would prefer moving the raise line above it into that block, but keeping the check for is_purged?. It is a guard clause that quickly skips already purged objects further SQL calls to get, say, a parent.